Apecs doesn't build with template-haskell-2.15.0.0

```bash
apecs  > [ 5 of 12] Compiling Apecs.THTuples
apecs  >
apecs  > /tmp/stack7315/apecs-0.8.2/src/Apecs/THTuples.hs:58:11: error:
apecs  >     • Couldn't match expected type ‘(Type -> TySynEqn) -> Dec’
apecs  >                   with actual type ‘Dec’
apecs  >     • The first argument of ($) takes one argument,
apecs  >       but its type ‘Dec’ has none
apecs  >       In the expression:
apecs  >         TySynInstD strgN
apecs  >           $ TySynEqn [varTuple] (tupleUpT . fmap strgT $ vars)
apecs  >       In the fourth argument of ‘InstanceD’, namely
apecs  >         ‘[TySynInstD strgN
apecs  >             $ TySynEqn [varTuple] (tupleUpT . fmap strgT $ vars)]’
apecs  >    |
apecs  > 58 |         [ TySynInstD strgN $
apecs  >    |           ^^^^^^^^^^^^^^^^^^...
apecs  >
apecs  > /tmp/stack7315/apecs-0.8.2/src/Apecs/THTuples.hs:58:22: error:
apecs  >     • Couldn't match expected type ‘TySynEqn’ with actual type ‘Name’
apecs  >     • In the first argument of ‘TySynInstD’, namely ‘strgN’
apecs  >       In the expression: TySynInstD strgN
apecs  >       In the expression:
apecs  >         TySynInstD strgN
apecs  >           $ TySynEqn [varTuple] (tupleUpT . fmap strgT $ vars)
apecs  >    |
apecs  > 58 |         [ TySynInstD strgN $
apecs  >    |                      ^^^^^
apecs  >
apecs  > /tmp/stack7315/apecs-0.8.2/src/Apecs/THTuples.hs:59:20: error:
apecs  >     • Couldn't match expected type ‘Maybe [TyVarBndr]’
apecs  >                   with actual type ‘[Type]’
apecs  >     • In the first argument of ‘TySynEqn’, namely ‘[varTuple]’
apecs  >       In the second argument of ‘($)’, namely
apecs  >         ‘TySynEqn [varTuple] (tupleUpT . fmap strgT $ vars)’
apecs  >       In the expression:
apecs  >         TySynInstD strgN
apecs  >           $ TySynEqn [varTuple] (tupleUpT . fmap strgT $ vars)
apecs  >    |
apecs  > 59 |           TySynEqn [varTuple] (tupleUpT . fmap strgT $ vars)
apecs  >    |                    ^^^^^^^^^^
apecs  >
apecs  > /tmp/stack7315/apecs-0.8.2/src/Apecs/THTuples.hs:82:15: error:
apecs  >     • Couldn't match expected type ‘(Type -> TySynEqn) -> t’
apecs  >                   with actual type ‘Dec’
apecs  >     • The first argument of ($) takes one argument,
apecs  >       but its type ‘Dec’ has none
apecs  >       In the expression:
apecs  >         TySynInstD elemN $ TySynEqn [varTuple] (tupleUpT $ fmap elemT vars)
apecs  >       In an equation for ‘elemI’:
apecs  >           elemI
apecs  >             = TySynInstD elemN
apecs  >                 $ TySynEqn [varTuple] (tupleUpT $ fmap elemT vars)
apecs  >     • Relevant bindings include
apecs  >         elemI :: t (bound at src/Apecs/THTuples.hs:82:7)
apecs  >    |
apecs  > 82 |       elemI = TySynInstD elemN $ TySynEqn [varTuple] (tupleUpT $ fmap elemT vars)
apecs  >    |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
apecs  >
apecs  > /tmp/stack7315/apecs-0.8.2/src/Apecs/THTuples.hs:82:26: error:
apecs  >     • Couldn't match expected type ‘TySynEqn’ with actual type ‘Name’
apecs  >     • In the first argument of ‘TySynInstD’, namely ‘elemN’
apecs  >       In the expression: TySynInstD elemN
apecs  >       In the expression:
apecs  >         TySynInstD elemN $ TySynEqn [varTuple] (tupleUpT $ fmap elemT vars)
apecs  >    |
apecs  > 82 |       elemI = TySynInstD elemN $ TySynEqn [varTuple] (tupleUpT $ fmap elemT vars)
apecs  >    |                          ^^^^^
apecs  >
apecs  > /tmp/stack7315/apecs-0.8.2/src/Apecs/THTuples.hs:82:43: error:
apecs  >     • Couldn't match expected type ‘Maybe [TyVarBndr]’
apecs  >                   with actual type ‘[Type]’
apecs  >     • In the first argument of ‘TySynEqn’, namely ‘[varTuple]’
apecs  >       In the second argument of ‘($)’, namely
apecs  >         ‘TySynEqn [varTuple] (tupleUpT $ fmap elemT vars)’
apecs  >       In the expression:
apecs  >         TySynInstD elemN $ TySynEqn [varTuple] (tupleUpT $ fmap elemT vars)
apecs  >    |
apecs  > 82 |       elemI = TySynInstD elemN $ TySynEqn [varTuple] (tupleUpT $ fmap elemT vars)
apecs  >    |                                           ^^^^^^^^^^
apecs  >

--  While building package apecs-0.8.2 using:
      /home/spacekitteh/.stack/setup-exe-cache/x86_64-linux/Cabal-simple_mPHDZzAJ_3.0.0.0_ghc-8.8.1 --builddir=.stack-work/dist/x86_64-linux/Cabal-3.0.0.0 build --ghc-options " -fdiagnostics-color=always"
    Process exited with code: ExitFailure 1```

--- a/.travis.yml
+++ b/.travis.yml
@@ -32,10 +32,6 @@ matrix:
     compiler: ": #stack default"
     addons: {apt: {packages: [libgmp-dev,freeglut3-dev]}}
 
-  - env: BUILD=stack ARGS="--resolver lts-12"
-    compiler: ": #stack 8.4.4"
-    addons: {apt: {packages: [libgmp-dev,freeglut3-dev]}}
-
   - env: BUILD=stack ARGS="--resolver lts-13"
     compiler: ": #stack 8.6.5"
     addons: {apt: {packages: [libgmp-dev,freeglut3-dev]}}
@@ -51,10 +47,6 @@ matrix:
 
   - env: BUILD=stack ARGS=""
     compiler: ": #stack default osx"
-    os: osx
-
-  - env: BUILD=stack ARGS="--resolver lts-12"
-    compiler: ": #stack 8.4.4 osx"
     os: osx
 
   - env: BUILD=stack ARGS="--resolver lts-13"

--- a/apecs/apecs.cabal
+++ b/apecs/apecs.cabal
@@ -1,5 +1,5 @@
 name:                apecs
-version:             0.8.2
+version:             0.8.3
 homepage:            https://github.com/jonascarpay/apecs#readme
 license:             BSD3
 license-file:        LICENSE

--- a/apecs/src/Apecs/THTuples.hs
+++ b/apecs/src/Apecs/THTuples.hs
@@ -1,4 +1,5 @@
 {-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE CPP #-}
 
 module Apecs.THTuples where
 
@@ -54,9 +55,12 @@ tupleInstances n = do
       compT var = ConT compN `AppT` var
       strgN = mkName "Storage"
       strgT var = ConT strgN `AppT` var
-      compI = InstanceD Nothing (fmap compT vars) (compT varTuple)
-        [ TySynInstD strgN $
-          TySynEqn [varTuple] (tupleUpT . fmap strgT $ vars)
+      compI = InstanceD Nothing (fmap compT vars) (compT varTuple) [
+#if MIN_VERSION_template_haskell(2,15,0)
+          TySynInstD $ TySynEqn Nothing (strgT varTuple) (tupleUpT . fmap strgT $ vars)
+#else
+          TySynInstD strgN $ TySynEqn [varTuple] (tupleUpT . fmap strgT $ vars)
+#endif
         ]
 
       -- Has
@@ -79,7 +83,11 @@ tupleInstances n = do
       -- Elem
       elemN = mkName "Elem"
       elemT var = ConT elemN `AppT` var
+#if MIN_VERSION_template_haskell(2,15,0)
+      elemI = TySynInstD $ TySynEqn Nothing (elemT varTuple) (tupleUpT $ fmap elemT vars)
+#else
       elemI = TySynInstD elemN $ TySynEqn [varTuple] (tupleUpT $ fmap elemT vars)
+#endif
 
       -- s, ety, w arguments
       sNs = [ mkName $ "s_" ++ show i | i <- [0..n-1]]

--- a/stack.yaml
+++ b/stack.yaml
@@ -11,10 +11,6 @@ extra-deps:
   - stm-containers-1.1.0.2
   - stm-hamt-1.2.0.2
   - primitive-extras-0.7.1
-  - deferred-folds-0.9.10
-  - focus-1.0.1.3
-  - primitive-0.6.4.0
-
 
 nix:
   pure: false

